### ISSUE TITLE
GLFW 3.2+: glfwInit() hangs for up to a minute

### ISSUE THREAD

**Issue by Xerolyph** [glfw/glfw#1284]
I have an issue where my program hangs on glfwInit() for any amount of time between 20 and 50 seconds or so; it varies each time i try to run the application.

I've tried using LWJGL 3.1.6 as well as directly compiling an application in C++ using GLFW, both of which yield the same result (LWJGL 3.1.6 uses GLFW pre-release 3.3).

I tried compiling a couple different versions of GLFW in C++. Compiling with 3.2 and 3.2.1 reproduces the issue, but using 3.1.2 does not reproduce it, and the program starts fine without any delay. I should note that no error messages are shown in the callback when reproducing the issue. There is no difference in the result between compiling for 32-bit or 64-bit.

GLFW version: 3.2+
OS: Windows 10 update 1803, 64-bit

The issue will occur when writing a simple window.

**Comment by OlivierSohn** [comment]
Do you have a minimal sample program reproducing the issue?

**Comment by Xerolyph** [comment]
The example code shown here will reproduce it: http://www.glfw.org/documentation.html

The process will hang on the glfwInit() call for up to a minute, then it will continue as normal. I'm not sure what the cause may be, but I know that the issue doesn't occur in GLFW 3.1.2, so something that was changed between 3.1 and 3.2 may have broken something.

**Comment by OlivierSohn** [comment]
One way to debug this is to add logs inside glfwInit() and see where it blocks, can you try that?

**Comment by Xerolyph** [comment]
Alright, I added several prints and it appears to hang here:
```
if (!_glfwPlatformInit())
{
    _glfwPlatformTerminate();
    return GLFW_FALSE;
}
```
This is the only piece of code that it appears to be stuck on.

**Comment by OlivierSohn** [comment]
I guess going deeper is the way to go. I'm not familiar with that code, but if you find the exact call that takes time, then we can google the function name, see if other people are encountering the issue, etc...

It's probably within _glfwInitJoysticksWin32(), as the code of _glfwInitTimerWin32() seems harmless to me.

**Comment by Xerolyph** [comment]
Yeah, I edited my post as it was indeed the culprit. I tracked the pause to this check, inside _glfwDetectJoystickConnectionWin32(void), in win32_joystick.c:
```
if (FAILED(IDirectInput8_EnumDevices(_glfw.win32.dinput8.api,
                                             DI8DEVCLASS_GAMECTRL,
                                             deviceCallback,
                                             NULL,
                                             DIEDFL_ALLDEVICES)))
```

**Comment by OlivierSohn** [comment]
Then, you can dig deeper by adding logs to the `deviceCallback` function :)

**Comment by Xerolyph** [comment]
Hmm.. the `deviceCallback` function wasn't called.

**Comment by OlivierSohn** [comment]
ok, If you don't have any connected joystick it's normal I guess.

I'll let others help out, I think you can rename the title of the issue to something more specific (maybe mentionning `IDirectInput8_EnumDevices` in the name)

**Comment by OlivierSohn** [comment]
I've found that that may be related: https://discourse.libsdl.org/t/win-10-initing-joystick-causes-45s-hang-in-sdl-dinput-joystickdetect/23293

**Comment by Xerolyph** [comment]
The issue described in that post is pretty much what happens. I looked at the source code of GLFW 3.1.2, and it looks like the joystick handling was changed, which introduced this issue.

**Comment by OlivierSohn** [comment]
Another interesting read: 
https://social.msdn.microsoft.com/Forums/sqlserver/en-US/db490151-b26b-42d0-8cf5-131ef05776da/directinput-function-idirectinput8enumdevices-bug-on-win10?forum=vcgeneral

Can you try changing `DIEDFL_ALLDEVICES ` to `DIEDFL_ATTACHEDONLY` and see if it is faster?

**Comment by Xerolyph** [comment]
I changed the parameter accordingly and there was no change in the result. Here's something really strange, though - I stumbled across this thread:
https://stackoverflow.com/questions/10967795/directinput8-enumdevices-sometimes-painfully-slow

The last post mentioned the K55 keyboard, which is the keyboard I have. I tried unplugging it, and the issue was immediately resolved. When plugging the keyboard back in, I can run the program a good deal of times before the issue comes back several minutes after plugging it back in. Reinstalling the driver does not resolve the issue.

I don't really have another keyboard to test this with, but he mentioned in the post that plugging in a new keyboard only temporarily fixed the issue anyway.

**Comment by OlivierSohn** [comment]
Ok, it doesn't look like there is much that can be done on glfw side, then...

**Comment by otoomey** [comment]
I am using the java bindings for GLFW on windows and I am getting the exact same problem. Yesterday my project started in under a second. Today it takes almost thirty seconds. The GLFW Error Callback doesn't log anything either. I have not plugged or unplugged any devices from my computer.

**Comment by mcsantiago** [comment]
Phew I am glad that I am not the only one experiencing this issue. The weird part is that this is a recent behavior that I've encountered. I have not changed the includes nor the libs. Strange...

**Comment by eyum** [comment]
I've also been experiencing this issue, and I also have a K55. Glad to know I'm not alone. I've noticed that on rebooting my PC my program starts up quickly for the first few times before it starts hanging on the glfwInit.

**Comment by Xerolyph** [comment]
Just a note I should have said a while ago: switching to another keyboard does solve the problem. There may be some sort of driver issue with the K55. I saw that there was an issue with another game and the K55 together, and in that case the game actually crashed.

**Comment by Ihysoal** [comment]
> I changed the parameter accordingly and there was no change in the result. Here's something really strange, though: https://stackoverflow.com/questions/10967795/directinput8-enumdevices-sometimes-painfully-slow
> 
> The last post mentioned the K55 keyboard, which is the keyboard I have. I tried unplugging it, and the issue was immediately resolved. When plugging the keyboard back in, I can run the program a number of times before the issue comes back several minutes after plugging it back in. Reinstalling the driver does not resolve the issue.

In my case after reading the stackoverflow answers, I suspect it to have been caused by iCUE, which was not detecting my keyboard for some reason. I replugged the keyboard, it popped up in iCUE, and gflwInit() ran instantly again.

It may have been triggered by an update prompt and/or multiple restarts. If I find out what exactly triggers the issue I'll update my post.

**Comment by matusnovak** [comment]
I can also confirm that having a Corsair K55 keyboard causes this issue. If I unplug my keyboard the `IDirectInput8_EnumDevices` instantly finishes and the GLFW window starts. But, this does not happen every time. Rebooting the PC will fix the issue temporarily, but the hanging starts after I have restarted the GLFW application (which happens a lot if I am debugging).

If anyone has a solution for this please let me know or I will be forced to buy a different keyboard.

**Comment by nlutsenko** [comment]
Also got hit by it, even though in our context, we actually don't care about joystick functionality whatsoever.

Might I propose to actually make the joystick functionality be opt-out within glfw, as a permanent mitigation for this sort of issues?

Regardless, whether it takes a little bit of time or extends for up to a minute (we see an avg of 57s on Windows on one of the test machines with some bizarre combination of drivers), I truly believe in "don't pay for what you don't use" principle.
In the current implementation the joystick support is loaded unconditionally, even though, a given app integrating glfw might completely not care (our case), or might want to optimize further and do the handling themselves.

**TL;DR Proposal** Make the joystick functionality optional at initialization time (default: on).

**Comment by WillerZ** [comment]
@elmindreda would a PR to do what @nlutsenko describes above in a form something like a `GLFW_SKIP_INPUT_INIT` hint be welcome?

**Comment by tombsar** [comment]
I'd suggest an init hint named  `GLFW_JOYSTICK_INPUT` (or similar) with a default value of `GLFW_TRUE`. Defaulting to false and/or on-demand gamepad detection can be considered for the next API version, but probably not in the 3.x series.

**Comment by elmindreda** [comment]
A quick test implementation of delay-initialized joysticks is now available in the `joystick-delay-init` branch. It's had a tiny bit of testing on all supported platforms.

Does this branch work for you? If so, would you prefer this or the init hint?

**Comment by WillerZ** [comment]
I haven't tested it, but I read the code. It would work for us because we do not call any joystick-related functions.

With a more-general hat on: as you have a gamepad/joystick separation you could also consider using Xinput as the backend for gamepads on Windows - it doesn't have this delay problem that DirectInput has, and it's a supported API from MS so we can complain about it to them if it does degrade.

**Comment by WillerZ** [comment]
We have no preference for either route - I suggested the init hint mechanism because we made that change to our copy of glfw to fix this for ourselves and it would have been easy for us to make it a PR.

**Comment by elmindreda** [comment]
> With a more-general hat on: as you have a gamepad/joystick separation you could also consider using Xinput as the backend for gamepads on Windows - it doesn't have this delay problem that DirectInput has, and it's a supported API from MS so we can complain about it to them if it does degrade.

We use DirectInput (and SDL_GameControllerDB) in addition to XInput because many gamepad-shaped controllers are not available via XInput.

> We have no preference for either route - I suggested the init hint mechanism because we made that change to our copy of glfw to fix this for ourselves and it would have been easy for us to make it a PR.

That makes sense, thanks.

**Comment by elmindreda** [comment]
Delayed initialization of joystick support has been merged as 782e6b6cefc9ac2a2b6779f9e658ab948809253b.

**Comment by nek0bit** [comment]
This is so weird! I am still in shock that my corsair K55 had *some* meaning behind this because I never would've blamed my keyboard (I had this problem for 2 years now). I was developing a program using GLFW but I have had this issue with programs that use GLFW (maybe) such as Dolphin emulator. I tried debugging personally and found glfwInit() being stuck, so I googled, know already that the issue I had with other programs was similar to this, and holy shit, Unplugging my keyboard and switching to an old one I never used fixed the problem... it appears that rebooting also fixed the problem with the keyboard, but only temporarily

**Comment by rayzchen** [comment]
How do I disable joystick initialization? I have this error too.

**Comment by tombsar** [comment]
@rayzchen Delayed joystick initialization hasn't made it into a release yet (should be in 3.4), but checking out and compiling the current master from github should help.

**Comment by MarkusMiller** [comment]
Hello, is there anything in the current version (3.3.8) that can be done to avoid this excessive init-time? I don't know, what exactly causes the issue to happen (got a K95) after a reboot, but I'm losing such a fucking huge amount of time rebooting after it occurred or waiting till the window opens, that it starts not to be a practical windowing system for this project anymore. This also happened on a laptop of mine a couple of years ago, which had no Corsair keyboard attached to it. Would be great if I could prevent this bloody function from being called somehow (don't need any joystick support in the current project)

**Comment by AndrejMitrovic** [comment]
For me specifically I found that when I disable the Corsair services GLFW takes a long time to run `glfwInit()`. Enabling the Corsair services fixes the long running time but only if the keyboard's latest firmware is installed. There's an option somewhere in the iCUE software to update the firmware.

**Comment by MarkusMiller** [comment]
Are you restarting the CorsairService? I actually identified my corsair headset as the cause of trouble, because unplugging it from the USB charge or plugging it in may cause the issue. Not always, though.

**Comment by AndrejMitrovic** [comment]
I shut it down completely through the Services app on Windows. There's a way to disable automatic startup of those Corsair services in there.

But I'm honestly not sure, it seems to be random at times. Sometimes shutting down those Corsair services fixes the problem, other times turning them on fixes the problem.

**Comment by Andersama** [comment]
Just ran into this issue, weirdly enough have a corsair keyboard. Thing is I recall having a different glfw install / version / build I was using, but I got irritated and switched to having vcpkg do the work for a new project.

**Comment by Cathodeus** [comment]
I suffer from the same issue too ... I will try to explain it to Corsair Support, i highly doubt that they will do anything ...

**Comment by tksuoran** [comment]
This is now happening to me as well. I have a few USB devices, some maybe a bit unusual.
- RollerMouse Red
- 3D Connexion Space Nagivator
- SteelSeries Apex Pro TKL keyboard
- Some random bluetooth adapter - plugged into the USB port in the Apex Pro TKL keyboard.
- Some cheap Corsair mouse

I have some non-glfw code (hidapi, for using Space Navigator) which _also_ triggers delay. Calls to `HidD_GetSerialNumberString()`, `HidD_GetManufacturerString()` and `HidD_GetProductString()` take several seconds each, then fail, at least for USB device vendor `1b1c` product `1b75`. The vendor is Corsair.

Some testing I did:
- There was a firmware update for the keyboard, I installed it, it did not help
- I unplugged the bluetooth adapter from the keyboard, but that did not fix the issue.
- I unplugged the Space Navigator, but that did not fix the issue.
- I unplugged the cheap Corsair mouse. The issue was gone.

So looks like some bad (Corsair?) devices can cause the issue.

**Comment by KatekovAnton** [comment]
Hi. I had the same problem now. Digging a little bit into it I found that the problem was somehow caused by my mouse - I guess my Steelseries mouse somehow affected the enumeration method and reconnecting it might just drop some state in the system so enumeration wont hit it anymore.

Reconnecting the mouse removed the delay and now `_glfwPlatformInit` finishes fast. Once fixed, I was not able to deliberately reproduce the hanging anymore.

It may appear again so if you have any clues what too look at - please let me know so I could try to examine it. If not - hope this info can help narrowing down the search of the root cause of the problem.

Edit: my glfw is
```
FetchContent_Declare(
        glfw
        GIT_REPOSITORY https://github.com/glfw/glfw.git
        GIT_TAG        dc46d3f8129712e42856c20e99a604a3b08ad581 # 3.3.10
    )
```
